File: src/workers/typescript/analyzer.js
const parser = require('@babel/parser')
const traverse = require('@babel/traverse').default
const {
  createAnalysisEnvelope,
  addEntity,
  addRelation,
  finalizeAnalysis
} = require('../../shared/analysis-schema')

function getFunctionName(path) {
  if (path.node.id?.name) return path.node.id.name

  const parent = path.parent
  if (parent?.type === 'VariableDeclarator' && parent.id?.name) return parent.id.name
  if (parent?.type === 'AssignmentExpression' && parent.left?.type === 'Identifier') return parent.left.name
  if (parent?.type === 'ObjectProperty' && parent.key?.name) return parent.key.name

  return `<anonymous:${path.node.start}>`
}

function getParamName(param) {
  if (!param) return '<unknown>'
  if (param.type === 'Identifier') return param.name
  if (param.type === 'AssignmentPattern' && param.left.type === 'Identifier') return param.left.name
  if (param.type === 'RestElement' && param.argument.type === 'Identifier') return `...${param.argument.name}`
  return '<complex-param>'
}

function getCalleeName(node) {
  if (!node) return '<unknown>'
  if (node.type === 'Identifier') return node.name
  if (node.type === 'MemberExpression') {
    const objectName = node.object?.name || '<expr>'
    const propertyName = node.property?.name || '<member>'
    return `${objectName}.${propertyName}`
  }
  if (node.type === 'Import') return 'import'
  return node.type || '<unknown>'
}

function getTypeParameterNames(typeParameters) {
  if (!typeParameters?.params?.length) return []
  return typeParameters.params.map(param => param.name || '<type-param>')
}

function getImportBindings(specifiers = []) {
  return specifiers.map(specifier => {
    if (specifier.type === 'ImportSpecifier') {
      return {
        imported: specifier.imported?.name || specifier.imported?.value || '<unknown>',
        local: specifier.local?.name || '<unknown>',
        kind: 'named'
      }
    }

    if (specifier.type === 'ImportDefaultSpecifier') {
      return {
        imported: 'default',
        local: specifier.local?.name || 'default',
        kind: 'default'
      }
    }

    if (specifier.type === 'ImportNamespaceSpecifier') {
      return {
        imported: '*',
        local: specifier.local?.name || '<namespace>',
        kind: 'namespace'
      }
    }

    return {
      imported: '<unknown>',
      local: '<unknown>',
      kind: 'unknown'
    }
  })
}

function analyzeTypeScript(code, metadata = {}) {
  const analysis = createAnalysisEnvelope('typescript', 'babel', metadata)
  const functionStack = []
  const moduleId = analysis.module.id

  const ast = parser.parse(code, {
    sourceType: 'unambiguous',
    errorRecovery: true,
    plugins: [
      'typescript',
      'jsx',
      'classProperties',
      'decorators-legacy',
      'dynamicImport'
    ]
  })

  analysis.diagnostics = ast.errors.map(error => ({
    severity: 'warning',
    message: error.message,
    position: error.loc || null
  }))

  traverse(ast, {
    FunctionDeclaration: {
      enter(path) {
        const name = getFunctionName(path)
        const entityId = `${moduleId}:function:${name}`
        const parentId = functionStack[functionStack.length - 1] || moduleId

        addEntity(analysis, {
          id: entityId,
          kind: 'function',
          name,
          parentId,
          data: {
            params: path.node.params.map(getParamName),
            runtimeKind: 'FunctionDeclaration',
            typeParameters: getTypeParameterNames(path.node.typeParameters)
          }
        })

        addRelation(analysis, {
          id: `${parentId}->${entityId}:contains`,
          from: parentId,
          to: entityId,
          kind: 'contains',
          data: {}
        })

        functionStack.push(entityId)
      },
      exit() {
        functionStack.pop()
      }
    },
    FunctionExpression: {
      enter(path) {
        const name = getFunctionName(path)
        const entityId = `${moduleId}:function:${name}`
        const parentId = functionStack[functionStack.length - 1] || moduleId

        addEntity(analysis, {
          id: entityId,
          kind: 'function',
          name,
          parentId,
          data: {
            params: path.node.params.map(getParamName),
            runtimeKind: 'FunctionExpression',
            typeParameters: getTypeParameterNames(path.node.typeParameters)
          }
        })

        addRelation(analysis, {
          id: `${parentId}->${entityId}:contains`,
          from: parentId,
          to: entityId,
          kind: 'contains',
          data: {}
        })

        functionStack.push(entityId)
      },
      exit() {
        functionStack.pop()
      }
    },
    ArrowFunctionExpression: {
      enter(path) {
        const name = getFunctionName(path)
        const entityId = `${moduleId}:function:${name}`
        const parentId = functionStack[functionStack.length - 1] || moduleId

        addEntity(analysis, {
          id: entityId,
          kind: 'function',
          name,
          parentId,
          data: {
            params: path.node.params.map(getParamName),
            runtimeKind: 'ArrowFunctionExpression',
            typeParameters: getTypeParameterNames(path.node.typeParameters)
          }
        })

        addRelation(analysis, {
          id: `${parentId}->${entityId}:contains`,
          from: parentId,
          to: entityId,
          kind: 'contains',
          data: {}
        })

        functionStack.push(entityId)
      },
      exit() {
        functionStack.pop()
      }
    },
    ClassDeclaration(path) {
      const name = path.node.id?.name || `<anonymous-class:${path.node.start}>`
      const entityId = `${moduleId}:class:${name}`

      addEntity(analysis, {
        id: entityId,
        kind: 'class',
        name,
        parentId: moduleId,
        data: {
          superClass: path.node.superClass?.name || null,
          typeParameters: getTypeParameterNames(path.node.typeParameters),
          implements: (path.node.implements || []).map(item => item.expression?.name || '<unknown>')
        }
      })

      addRelation(analysis, {
        id: `${moduleId}->${entityId}:contains`,
        from: moduleId,
        to: entityId,
        kind: 'contains',
        data: {}
      })
    },
    TSInterfaceDeclaration(path) {
      const name = path.node.id.name
      const entityId = `${moduleId}:interface:${name}`

      addEntity(analysis, {
        id: entityId,
        kind: 'interface',
        name,
        parentId: moduleId,
        data: {
          extends: (path.node.extends || []).map(item => item.expression?.name || '<unknown>'),
          typeParameters: getTypeParameterNames(path.node.typeParameters)
        }
      })

      addRelation(analysis, {
        id: `${moduleId}->${entityId}:contains`,
        from: moduleId,
        to: entityId,
        kind: 'contains',
        data: {}
      })
    },
    TSTypeAliasDeclaration(path) {
      const name = path.node.id.name
      const entityId = `${moduleId}:type:${name}`

      addEntity(analysis, {
        id: entityId,
        kind: 'type',
        name,
        parentId: moduleId,
        data: {
          typeParameters: getTypeParameterNames(path.node.typeParameters),
          targetType: path.node.typeAnnotation?.type || '<unknown>'
        }
      })

      addRelation(analysis, {
        id: `${moduleId}->${entityId}:contains`,
        from: moduleId,
        to: entityId,
        kind: 'contains',
        data: {}
      })
    },
    TSEnumDeclaration(path) {
      const name = path.node.id.name
      const entityId = `${moduleId}:enum:${name}`

      addEntity(analysis, {
        id: entityId,
        kind: 'enum',
        name,
        parentId: moduleId,
        data: {
          members: path.node.members.map(member => member.id?.name || member.id?.value || '<unknown>')
        }
      })

      addRelation(analysis, {
        id: `${moduleId}->${entityId}:contains`,
        from: moduleId,
        to: entityId,
        kind: 'contains',
        data: {}
      })
    },
    ImportDeclaration(path) {
      addRelation(analysis, {
        id: `${moduleId}->import:${path.node.source.value}`,
        from: moduleId,
        to: `external:${path.node.source.value}`,
        kind: 'imports',
        data: {
          specifier: path.node.source.value,
          bindings: getImportBindings(path.node.specifiers)
        }
      })
    },
    VariableDeclarator(path) {
      if (path.node.id.type !== 'Identifier') return

      const entityId = `${moduleId}:variable:${path.node.id.name}:${path.node.start}`
      const parentId = functionStack[functionStack.length - 1] || moduleId

      addEntity(analysis, {
        id: entityId,
        kind: 'variable',
        name: path.node.id.name,
        parentId,
        data: {
          runtimeKind: 'VariableDeclarator'
        }
      })

      addRelation(analysis, {
        id: `${parentId}->${entityId}:contains`,
        from: parentId,
        to: entityId,
        kind: 'contains',
        data: {}
      })
    },
    CallExpression(path) {
      const currentFunctionId = functionStack[functionStack.length - 1] || null
      if (!currentFunctionId) return

      const calleeName = getCalleeName(path.node.callee)

      addRelation(analysis, {
        id: `${currentFunctionId}->call:${calleeName}:${path.node.start}`,
        from: currentFunctionId,
        to: `call:${calleeName}`,
        kind: 'calls',
        data: {
          callee: calleeName
        }
      })
    }
  })

  return finalizeAnalysis(analysis)
}

module.exports = {
  analyzeTypeScript
}

File: src/workers/python/analyzer.py
import ast


def _node_name(node):
    if node is None:
        return "<unknown>"
    if isinstance(node, ast.Name):
        return node.id
    if isinstance(node, ast.Attribute):
        return f"{_node_name(node.value)}.{node.attr}"
    if isinstance(node, ast.Call):
        return _node_name(node.func)
    return node.__class__.__name__


def create_analysis_envelope(metadata):
    module_id = f"module:{metadata.get('objectKey') or metadata.get('originalName') or metadata.get('hash') or 'unknown'}"
    module_name = metadata.get("originalName") or metadata.get("objectKey") or "unknown"

    return {
        "schemaVersion": "2.0.0",
        "language": "python",
        "parser": "python-ast",
        "module": {
            "id": module_id,
            "name": module_name,
            "path": module_name,
            "hash": metadata.get("hash"),
            "language": "python",
        },
        "summary": {
            "modules": 1,
            "entities": 0,
            "relations": 0,
            "imports": 0,
            "calls": 0,
            "diagnostics": 0,
        },
        "entities": [
            {
                "id": module_id,
                "kind": "module",
                "name": module_name,
                "parentId": None,
                "data": {
                    "path": module_name,
                    "hash": metadata.get("hash"),
                },
            }
        ],
        "relations": [],
        "diagnostics": [],
    }


def finalize_analysis(analysis):
    analysis["summary"] = {
        "modules": 1,
        "entities": len(analysis["entities"]),
        "relations": len(analysis["relations"]),
        "imports": len([relation for relation in analysis["relations"] if relation["kind"] == "imports"]),
        "calls": len([relation for relation in analysis["relations"] if relation["kind"] == "calls"]),
        "diagnostics": len(analysis["diagnostics"]),
    }
    return analysis


class PythonAnalyzer(ast.NodeVisitor):
    def __init__(self, metadata):
        self.analysis = create_analysis_envelope(metadata)
        self.module_id = self.analysis["module"]["id"]
        self.function_stack = []

    def _add_entity(self, entity):
        self.analysis["entities"].append(entity)

    def _add_relation(self, relation):
        self.analysis["relations"].append(relation)

    def _register_function(self, name, kind, args):
        entity_id = f"{self.module_id}:function:{name}"
        parent_id = self.function_stack[-1] if self.function_stack else self.module_id

        self._add_entity({
            "id": entity_id,
            "kind": "function",
            "name": name,
            "parentId": parent_id,
            "data": {
                "runtimeKind": kind,
                "params": args,
            },
        })

        self._add_relation({
            "id": f"{parent_id}->{entity_id}:contains",
            "from": parent_id,
            "to": entity_id,
            "kind": "contains",
            "data": {},
        })

        self.function_stack.append(entity_id)

    def visit_FunctionDef(self, node):
        self._register_function(node.name, "FunctionDef", [arg.arg for arg in node.args.args])
        self.generic_visit(node)
        self.function_stack.pop()

    def visit_AsyncFunctionDef(self, node):
        self._register_function(node.name, "AsyncFunctionDef", [arg.arg for arg in node.args.args])
        self.generic_visit(node)
        self.function_stack.pop()

    def visit_ClassDef(self, node):
        entity_id = f"{self.module_id}:class:{node.name}"
        self._add_entity({
            "id": entity_id,
            "kind": "class",
            "name": node.name,
            "parentId": self.module_id,
            "data": {
                "bases": [_node_name(base) for base in node.bases],
                "methods": [
                    child.name
                    for child in node.body
                    if isinstance(child, (ast.FunctionDef, ast.AsyncFunctionDef))
                ],
            },
        })

        self._add_relation({
            "id": f"{self.module_id}->{entity_id}:contains",
            "from": self.module_id,
            "to": entity_id,
            "kind": "contains",
            "data": {},
        })

        self.generic_visit(node)

    def visit_Import(self, node):
        for alias in node.names:
            self._add_relation({
                "id": f"{self.module_id}->import:{alias.name}",
                "from": self.module_id,
                "to": f"external:{alias.name}",
                "kind": "imports",
                "data": {
                    "specifier": alias.name,
                    "bindings": [{
                        "imported": alias.name,
                        "local": alias.asname or alias.name,
                        "kind": "module",
                    }],
                },
            })

    def visit_ImportFrom(self, node):
        module_name = node.module or "<relative>"
        self._add_relation({
            "id": f"{self.module_id}->import:{module_name}",
            "from": self.module_id,
            "to": f"external:{module_name}",
            "kind": "imports",
            "data": {
                "specifier": module_name,
                "bindings": [
                    {
                        "imported": alias.name,
                        "local": alias.asname or alias.name,
                        "kind": "named",
                    }
                    for alias in node.names
                ],
            },
        })

    def visit_Assign(self, node):
        for target in node.targets:
            if isinstance(target, ast.Name):
                entity_id = f"{self.module_id}:variable:{target.id}:{node.lineno}"
                parent_id = self.function_stack[-1] if self.function_stack else self.module_id
                self._add_entity({
                    "id": entity_id,
                    "kind": "variable",
                    "name": target.id,
                    "parentId": parent_id,
                    "data": {
                        "runtimeKind": "Assign",
                    },
                })
                self._add_relation({
                    "id": f"{parent_id}->{entity_id}:contains",
                    "from": parent_id,
                    "to": entity_id,
                    "kind": "contains",
                    "data": {},
                })
        self.generic_visit(node)

    def visit_AnnAssign(self, node):
        if isinstance(node.target, ast.Name):
            entity_id = f"{self.module_id}:variable:{node.target.id}:{node.lineno}"
            parent_id = self.function_stack[-1] if self.function_stack else self.module_id
            self._add_entity({
                "id": entity_id,
                "kind": "variable",
                "name": node.target.id,
                "parentId": parent_id,
                "data": {
                    "runtimeKind": "AnnAssign",
                },
            })
            self._add_relation({
                "id": f"{parent_id}->{entity_id}:contains",
                "from": parent_id,
                "to": entity_id,
                "kind": "contains",
                "data": {},
            })
        self.generic_visit(node)

    def visit_Call(self, node):
        current_function = self.function_stack[-1] if self.function_stack else None
        if current_function:
            callee = _node_name(node.func)
            self._add_relation({
                "id": f"{current_function}->call:{callee}:{node.lineno}",
                "from": current_function,
                "to": f"call:{callee}",
                "kind": "calls",
                "data": {
                    "callee": callee,
                },
            })
        self.generic_visit(node)


def analyze_python(code, metadata=None):
    metadata = metadata or {}

    try:
        tree = ast.parse(code)
    except SyntaxError as error:
        analysis = create_analysis_envelope(metadata)
        analysis["diagnostics"] = [{
            "severity": "error",
            "message": str(error),
            "position": {
                "line": error.lineno,
                "column": error.offset,
            },
        }]
        return finalize_analysis(analysis)

    analyzer = PythonAnalyzer(metadata)
    analyzer.visit(tree)
    return finalize_analysis(analyzer.analysis)
